src/bucket/BucketManagerImpl.cpp

```diff
@@ -656,24 +698,41 @@ BucketManagerImpl::forgetUnreferencedBuckets()
 {
     ZoneScoped;
     std::lock_guard<std::recursive_mutex> lock(mBucketMutex);
-    auto referenced = getReferencedBuckets();
+    auto referenced = getAllReferencedBuckets();
+    auto blReferenced = getBucketListReferencedBuckets();
 
     for (auto i = mSharedBuckets.begin(); i != mSharedBuckets.end();)
     {
-        // Standard says map iterators other than the one you're erasing remain
-        // valid.
+        // Standard says map iterators other than the one you're erasing
+        // remain valid.
         auto j = i;
         ++i;
 
+        // Delete indexes for buckets no longer in bucketlist. There is a race
+        // condition on startup where future buckets for a level will be
+        // finished and have an index but will not yet be refered to by the
+        // bucket level's next pointer. Checking use_count == 1 makes sure no
+        // other in-progress structures will add bucket to bucket list after
+        // deleting index
+        if (j->second->isIndexed() && j->second.use_count() == 1 &&
+            blReferenced.find(j->first) == blReferenced.end())
+        {
+            CLOG_TRACE(Bucket,
+                       "BucketManager::forgetUnreferencedBuckets deleting "
+                       "index for {}",
+                       j->second->getFilename());
+            j->second->freeIndex();
```

Comment:
Is there really any need for a separate method `freeIndex` at all? I mean .. would it not be adequate to have the bucket keep its index alive as long as the bucket is alive? Do we have buckets that get an index, and then we delete the index, but then also somehow keep the bucket alive?

Reply:
This is to free memory for buckets that are no longer referenced by the `BucketList` (i.e., will never be searched again). These are buckets referenced by the LCL and the publish queue. There's no reason for these buckets to sit around in the queue with (potentially very large) indexes. This is also why we split `getReferencedBuckets` into `getBucketListReferencedBuckets` and `getAllReferencedBuckets`. Buckets that are referenced by the BucketList are not eligible to free their index, while the other referenced buckets are eligible.

Reply:
ok